src/cargo/ops/cargo_new.rs

```diff
@@ -56,12 +56,33 @@ pub fn new(opts: NewOptions, config: &Config) -> CargoResult<()> {
         return Err(human(&format!("Invalid character `{}` in crate name: `{}`",
                                   c, name)));
     }
-    mk(config, &path, name, &opts).chain_error(|| {
+    let name =
+        if opts.bin {
+            name.to_string()
```

Comment:
all this string copying is not necessary, you ought to be able to continue to use the same underlying buffer